fendermint/actors/eam/src/state.rs

```diff
@@ -0,0 +1,116 @@
+use fvm_ipld_encoding::tuple::*;
+use fvm_shared::address::Address;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+pub type DeployerMap<BS> = Map2<BS, Address, ()>;
+
+/// The permission mode for controlling who can deploy contracts
+#[derive(Debug, Clone, PartialEq)]
+enum PermissionMode {
+    /// No restriction, everyone can deploy
+    NoRestriction,
+    /// Only whitelisted addresses can deploy
+    Whitelist(Cid), // HAMT[Address]()
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
+pub struct State {
+    permission_mode: PermissionMode,
+}
+
+impl State {
+    pub fn new<BS: Blockstore>(store: &BS, deployers: Vec<Address>) -> Result<State, ActorError> {
+        let mode = if deployers.is_empty() {
+            PermissionMode::NoRestriction
+        } else {
+            let mut deployers_map = DeployerMap::empty(store, DEFAULT_HAMT_CONFIG, "empty");
+            for d in deployers {
+                deployers_map.set(&d, ())?;
+            }
+            PermissionMode::Whitelist(deployers_map.flush()?)
+        };
+        Ok(State {
+            permission_mode: mode,
+        })
+    }
+
+    pub fn can_deploy(
+        &self,
+        store: &impl Blockstore,
+        deployer: &Address,
+    ) -> Result<bool, ActorError> {
+        Ok(match &self.permission_mode {
+            PermissionMode::NoRestriction => true,
+            PermissionMode::Whitelist(cid) => {
+                let deployers = DeployerMap::load(store, cid, DEFAULT_HAMT_CONFIG, "verifiers")?;
+                deployers.contains_key(deployer)?
+            }
+        })
+    }
+}
+
+const NO_RESTRICTION_MODE: u8 = 0;
+const WHITELIST_MODE: u8 = 1;
+
+impl Serialize for PermissionMode {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            PermissionMode::NoRestriction => {
+                let inner: (_, Vec<u8>) = (NO_RESTRICTION_MODE, vec![]);
+                Serialize::serialize(&inner, serde_tuple::Serializer(serializer))
+            }
+            PermissionMode::Whitelist(cid) => {
+                let inner = (WHITELIST_MODE, cid.to_bytes());
+                Serialize::serialize(&inner, serde_tuple::Serializer(serializer))
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for PermissionMode {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let (mode, bytes): (u8, Vec<u8>) =
+            Deserialize::deserialize(serde_tuple::Deserializer(deserializer))?;
+        Ok(match mode {
+            NO_RESTRICTION_MODE => PermissionMode::NoRestriction,
+            WHITELIST_MODE => PermissionMode::Whitelist(
+                Cid::from_bytes(&bytes).map_err(|_| Error::custom("invalid cid"))?,
+            ),
+            _ => return Err(Error::custom("invalid permission mode")),
+        })
+    }
+}
```

Comment:
You should not be needing the custom serde here! According to @vmx [here](https://filecoinproject.slack.com/archives/C015KQQLQQ1/p1707317303298409), libipld_serde_cbor supports externally tagged enums (not clear to me how it lays them out though), as well as `Option` types. We might need to add the #[repr] tag so serde knows which type to use for the discriminator. Perhaps that's why it didn't work for you originally?